Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D4ILLReduction.py

```diff
@@ -0,0 +1,466 @@
+            GroupWorkspaces(InputWorkspaces=corrected_list, OutputWorkspace=corrected_ws)
+            if self.getProperty('ClearCache').value:
+                DeleteWorkspace(Workspace=ws)
+            return corrected_ws
+        correction_ws = self._create_dead_time_correction(ws, tau)
+        corrected_ws = '{}_dt_corrected'.format(ws)
+        Multiply(LHSWorkspace=ws, RHSWorkspace=correction_ws, OutputWorkspace=corrected_ws)
+        if self.getProperty('ClearCache').value:
+            DeleteWorkspace(Workspace=correction_ws)
+        return corrected_ws
+
+    def _create_efficiency_correction(self, efficiency_file_name):
+        """
+        Creates a workspace with efficiency factors for each detector.
+
+        Args:
+        efficiency_file_name: (str) path to the file containing efficiency correction factors for each detector
+        return: (str) name of the workspace containing efficiency correction
+        """
+        y_values = []
+        det_no = 0
+        with open(efficiency_file_name, "r") as efficiency:
+            for line in efficiency:
+                if "#" in line:
+                    continue
+                line = line.strip(' ')
+                columns = line.split()
+                if columns:
+                    corr_val = float(columns[2])
+                    y_values.append(corr_val if corr_val >= 0 else np.nan)
+                    det_no += 1
+        eff_corr_ws = 'eff_corr_ws'
+        x_values = np.array(0)
+        CreateWorkspace(DataX=x_values, DataY=np.array(y_values), NSpec=len(y_values), OutputWorkspace=eff_corr_ws)
+        return eff_corr_ws
+
+    def _correct_relative_efficiency(self, ws):
+        """
+        Corrects the relative efficiency of each detector using values provided in an external ASCII file.
+
+        Args:
+        ws: (str) name of the workspace to be corrected
+        return: (str) name of corrected workspace
+        """
+        if self.getProperty('EfficiencyCalibrationFile').isDefault:
+            return ws
+        calibration_file = self.getPropertyValue('EfficiencyCalibrationFile')
+        eff_corr_ws = self._create_efficiency_correction(calibration_file)
+
+        output_ws = '{}_efficiency_corrected'.format(ws)
+        Divide(LHSWorkspace=ws, RHSWorkspace=eff_corr_ws, OutputWorkspace=output_ws)
+        MaskDetectorsIf(InputWorkspace=output_ws, OutputWorkspace=output_ws, Operator='NotFinite')
+        if self.getProperty('ClearCache').value:
+            DeleteWorkspaces(WorkspaceList=[eff_corr_ws, ws])
+        return output_ws
+
+    def _convert_to_q(self, ws, output_ws):
+        """
+        Converts the bin axis values of the  provided workspace from scattering angle to momentum exchange.
+
+        Args:
+        ws: (str) name of the workspace with x-axis in 2theta units
+        return: (str) name of the workspace converted to q
+        """
+        Ei = mtd[ws].getRun().getLogData('Ei').value * 1.60218e-22  # in J, meV->J
```

Comment:
Could the conversion factor to Joules be placed in a header file ? By the way, is there a common place in Mantid to store all conversion factors ?